src/modules/connections/services/ConnectionService.ts

```diff
@@ -418,17 +446,88 @@ export class ConnectionService {
     return this.connectionRepository.getSingleByQuery({ threadId })
   }
 
+  private async getRouting(mediationRecord: MediationRecord | undefined, routingKeys: string[], my_endpoint?: string) {
+    let endpoint
+    if (mediationRecord) {
+      routingKeys = [...routingKeys, ...mediationRecord.routingKeys]
+      endpoint = mediationRecord.endpoint
+    }
+    // Create and store new key
+    const did_data = await this.wallet.createDid()
+    if (mediationRecord) {
+      // new did has been created and mediator needs to be updated with the public key.
+      mediationRecord = await this.keylistUpdatdAndAwait(mediationRecord, did_data[1])
+    } else {
+      // TODO: register recipient keys for relay
+      // TODO: check that recipient keys are in wallet
+    }
+    endpoint = endpoint ?? my_endpoint ?? this.config.getEndpoint()
+    const result = { mediationRecord, endpoint, routingKeys, did: did_data[0], verkey: did_data[1] }
+    return result
+  }
+
+  public async keylistUpdatdAndAwait(
+    mediationRecord: MediationRecord,
+    verKey: string,
+    timeout = 15000 // TODO: this should be a configurable value in agent config
+  ): Promise<MediationRecord> {
+    const message = this.createKeylistUpdateMessage(verKey)
+    const connection = await this.getById(mediationRecord.connectionId)
+
+    const sendUpdateKeylist = async () => {
+      message.setReturnRouting(ReturnRouteTypes.all) // return message on request response
+      const outboundMessage = createOutboundMessage(connection, message)
+      await this.messageSender.sendMessage(outboundMessage)
+    }
+    const condition = async (event: KeylistUpdatedEvent) => {
+      return mediationRecord.id === event.payload.mediationRecord.id
+    }
+    const results = await waitForEvent(
+      sendUpdateKeylist,
+      RoutingEventTypes.RecipientKeylistUpdated,
+      condition,
+      timeout,
+      this.eventEmitter
+    )
+    return (results as KeylistUpdatedEvent).payload.mediationRecord
+  }
+
+  public createKeylistUpdateMessage(verkey: Verkey): KeylistUpdateMessage {
+    const keylistUpdateMessage = new KeylistUpdateMessage({
+      updates: [
+        new KeylistUpdate({
+          action: KeylistUpdateAction.add,
+          recipientKey: verkey,
+        }),
+      ],
+    })
+    return keylistUpdateMessage
+  }
```

Comment:
I think this belongs more in one of the mediation services

Reply:
yes. I agree. I am in the process of debugging a bug and then I will move this code into mediation services. I originally placed it into connectionModule to prevent the need for dependency injection, but I think I can move it easily now.